### PR TITLE
refactor!: rework feature components for easier composition

### PR DESCRIPTION
**Pull request by porcellus**
## Summary of change
Extracting logic of feature components into custom hooks

## Related issues

-   https://github.com/supertokens/supertokens-core/issues/330 this will make the react implementation easier/cleaner

## Test Plan

Should cause no change in behavior

## Documentation changes

Should cause no change in behavior

## Checklist for important updates

-   [x] Changelog has been updated
-   [x] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [x] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [x] If added a new recipe interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [x] If I added a new recipe, I also added the recipe entry point into the `size-limit` section of `package.json` with the size limit set to the current size rounded up.

## TODO
- [x] Update internal contributing guide with new principles that are added
- [x] Update demo apps in this repo to use new version and any code changes that might be required for them (some of them use JS to manipulate the react UI, so that might need to change)
- [x] Since this is a breaking change, we should also solve this bug in this release: https://github.com/supertokens/supertokens-auth-react/issues/380
- [x] Component override for ThirdPartyForm was showing as Anonymous.
- [x] Update documentation about overriding
- [ ] Update all demo apps

### CODE REVIEW

**Comment by rishabhpoddar** [comment]
It's really hard to review this on a code level. I'd appreciate comments in the code highlighting how the new structure works on a high level. If not in the code, please add to the contributing docs in the relevant place. Thanks

**Comment by porcellus** [comment]
I'll write a quick summary here first and then work on adding them to the contributions doc, I think this belongs there instead of in one of the recipes.
So the main changes:
- `GeneralError` moved out of `FormBase` (possibly breaking interface change for overrides)
- Feature component logic is extracted to custom hooks we can reuse for easier composition
- Using extracted hooks for `thirdpartyemailpassword` instead of embedding through the feature comps

File: lib/ts/recipe/emailpassword/components/features/signInAndUp/index.tsx [hunk 1]
```diff
@@ -19,228 +19,276 @@
...
-        this.state = {
+export const useFeatureReducer = (recipe: Recipe | undefined) => {
+    return React.useReducer(
+        (oldState: SignInAndUpState, action: EmailPasswordSignInAndUpAction) => {
+            switch (action.type) {
+                case "setSignIn":
+                    return {
+                        ...oldState,
+                        isSignUp: false,
+                    };
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/components/features/signInAndUp/index.tsx:49]: is this missing a break;?

File: lib/ts/recipe/emailpassword/types.ts [hunk 2]
```diff
@@ -425,13 +390,13 @@ export type FormBaseProps<T> = {
...
-    error?: string;
-
     validateOnBlur?: boolean;
 
+    clearError: () => void;
+    onError: (error: string) => void;
     onSuccess?: (result: T & { status: "OK" }) => void;
 
-    callAPI: (fields: APIFormField[]) => Promise<FormBaseAPIResponse<T>>;
+    callAPI: (fields: APIFormField[], setValue: (id: string, value: string) => void) => Promise<FormBaseAPIResponse<T>>;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/types.ts:399]: What is the purpose of `setValue` function?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/emailpassword/types.ts (hunk 2 above)
Comment (porcellus) [reply lib/ts/recipe/emailpassword/types.ts:399]: It's used in `emailOrPhoneForm`: we can update the stored value after submitting i.e., guess the international number.

File: lib/ts/recipe/emailpassword/components/features/signInAndUp/index.tsx (hunk 1 above)
Comment (porcellus) [reply lib/ts/recipe/emailpassword/components/features/signInAndUp/index.tsx:49]: The return statements return from the entire function, making breaks not necessary.

**Comment by porcellus** [comment]
#380 should be solved since the referenced component was removed.

**Comment by rishabhpoddar** [comment]
# Feedback
- [ ] Title and sub title missing in reset password forms (both the forms)

**Pull request merged**